pteranodon/autoconnect.py

```diff
@@ -0,0 +1,246 @@
+
+        all_ports = range(self._port_range[0], self._port_range[1] + 1)
+
+        new_loop = asyncio.new_event_loop()
+        ports_future = asyncio.gather(
+            *[self._test_port_open(port, socket_kind) for port in all_ports]
+        )
+        port_data = new_loop.run_until_complete(ports_future)
+
+        port_data = [
+            port_tup for port_tup in port_data if port_tup is not None
+        ]  # Remove None tuples
+        open_ports = [
+            port_tup[0] for port_tup in port_data if port_tup[1]
+        ]  # Filter only open ports
+        new_loop.stop()
+
+        self._available_ports[socket_kind] = open_ports
+        return open_ports
+
+    async def _test_server(
+        self,
+        serial_dev: Optional[str] = None,
+        port: Optional[int] = None,
+        is_tcp: Optional[bool] = None,
+        conn_timeout: float = 1.0,
+    ) -> Optional[Dict[str, Any]]:
+        """
+        Tests the given server (or serial device) and returns the input arguments if a MAV SDK server is present.
+
+        :param serial_dev: The path to the serial device to test (None if not testing serial)
+        :type serial_dev: Optional[str]
+        :param port: The UDP or TCP port to test (None if not testing TCP or UDP)
+        :type port: Optional[int]
+        :param is_tcp: True if testing TCP, False if testing UDP, None if testing serial
+        :type is_tcp: Optional[bool]
+        :param conn_timeout: Timeout for heartbeat message wait
+        :type conn_timeout: float
+        :return: None if no server at path, a dictionary of parameters otherwise
+        :rtype: Optional[Dict[str, Any]]
+        """
+        if serial_dev is None and port is None and is_tcp is None:
+            return None
+
+        conn: Optional[mavutil.mavfile] = None
+
+        if serial_dev is not None:
+            conn = mavutil.mavlink_connection(device=serial_dev, baud=115200)
+
+        if port is not None:
+            prefix = "tcp" if is_tcp else "udp"
+            addr = f"{prefix}:{self._server_addr}:{port}"
+            conn = mavutil.mavlink_connection(addr)
+
+        # Invalid parameters
+        if conn is None:
+            return None
+
+        msg = conn.recv_match(type="HEARTBEAT", blocking=True, timeout=conn_timeout)
+
+        # Msg is None on timeout
+        if msg is None:
+            return None
+
+        return {
+            "serial_dev": serial_dev,
+            "port": port,
+            "is_tcp": is_tcp,
+        }
+
+    # For the given dictionary, get the corresponding string server path
+    def _get_server_value(self, data_dict: Optional[Dict[str, Any]]) -> Optional[str]:
+        if data_dict is None:
+            return None
+
+        serial_dev: Optional[str] = data_dict["serial_dev"]
+        port: Optional[int] = data_dict["port"]
+
+        if serial_dev is not None:
+            return serial_dev
+
+        if data_dict["is_tcp"]:
+            return f"tcp://{self._server_addr}:{port}"
```

Comment:
Seems good to me! MAVSDK can accept a string without the address part, but it gets replaced with 127.0.0.1, so as long as this uses 127.0.0.1 or nothing then it is good. 

Has any of this been tested?